From 1809e0b737a3e798180f839d71ca49c75ff64576 Mon Sep 17 00:00:00 2001
From: Suraj Gupta <suraj.gupta2@amd.com>
Date: Wed, 25 Jun 2025 12:38:38 +0530
Subject: [PATCH 106/110] net: axienet: Get MRMAC GT lanes count based on port
 speed

Number of GT lanes for MRMAC depends on configured port speed.
Add get_mrmac_gtlanes_count() to get correct GT lane count.

Signed-off-by: Suraj Gupta <suraj.gupta2@amd.com>
Reviewed-by: Harini Katakam <harini.katakam@amd.com>
Message-ID: <20250625070839.2473420-2-suraj.gupta2@amd.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet.h  |  1 -
 .../net/ethernet/xilinx/xilinx_axienet_main.c | 25 +++++++++++++++++--
 2 files changed, 23 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet.h b/drivers/net/ethernet/xilinx/xilinx_axienet.h
index f9a7e79293aae..f61b1abb858b9 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet.h
@@ -732,7 +732,6 @@ enum temac_stat {
 #define MRMAC_GT_25G_MASK		0x00000002
 
 #define MRMAC_GT_LANE_OFFSET		BIT(16)
-#define MRMAC_MAX_GT_LANES		4
 #define GT_MODE_NARROW			"Narrow"
 
 /* DCMAC Register Definitions */
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index 5576020391b3c..fcff18332f514 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -835,14 +835,35 @@ static int dcmac_gt_reset(struct net_device *ndev)
 	return ret;
 }
 
+static inline int get_mrmac_gtlanes_count(int port_speed)
+{
+	int lane_count;
+
+	switch (port_speed) {
+	case SPEED_40000:
+	case SPEED_100000:
+		lane_count = 1;
+		break;
+	case SPEED_50000:
+		lane_count = 2;
+		break;
+	default:
+		/* 10G / 25G */
+		lane_count = 4;
+		break;
+	}
+	return lane_count;
+}
+
 static inline int axienet_mrmac_gt_reset(struct net_device *ndev)
 {
 	struct axienet_local *lp = netdev_priv(ndev);
 	u32 err, val;
-	int i;
+	int i, num_gtlanes;
 
+	num_gtlanes = get_mrmac_gtlanes_count(lp->max_speed);
 	if (mrmac_pll_rst == 0) {
-		for (i = 0; i < MRMAC_MAX_GT_LANES; i++) {
+		for (i = 0; i < num_gtlanes; i++) {
 			iowrite32(MRMAC_GT_RST_ALL_MASK, (lp->gt_ctrl +
 				  (MRMAC_GT_LANE_OFFSET * i) +
 				  MRMAC_GT_CTRL_OFFSET));
-- 
2.43.0

